api/src/middleware/auth.ts

```diff
@@ -0,0 +1,39 @@
+import { Request, Response, NextFunction } from 'express';
+import { hasRole, allRoles } from '../utils/auth-utils';
+
+/**
+ * Middleware to prevent users that aren't registered or don't have a
+ * specified role from completing requests
+ *
+ * @param roles the set of roles allowed to make the given request
+ * @returns 401 HTTP Response if the user lacks the necessary role, else calls
+ *          next() to continue to request
+ */
+const requireRole = (roles: typeof allRoles) => (
+  req: Request,
+  res: Response,
+  next: NextFunction,
+) => {
+  if (req.isUnauthenticated() || !hasRole(req.user, roles)) {
+    return res.status(401).json({
+      success: false,
+      message: 'Unauthorized',
+    });
+  }
+
+  next();
+};
+
+// Requires a user to have an account
+const requireRegistered = requireRole(allRoles);
```

Comment:
It might be easier to just use the exact constant from `rolesEnum` here instead of extracting it into another variable - that way you don't need to comment above each line and use arbitrary slicing

Reply:
The reason for this was because Admin should be allowed to stuff that Staff and contributors also have access to so it keeps the proper flow instead having to say isContributor or isStaff or isAdmin for Contributor permissions. I modeled this off of the member DB with the way Jeffrey did, let me know if you think this should still be changed

Reply:
Ohh ok I see. What if you just listed each emum explicitly like i.e. `rolesEnum.CONTRIBUTOR, rolesEnum.STAFF, rolesEnum.ADMIN` for `requireAdmin` so you can see the exact role structure immediately instead of having to find where `allRoles` is defined?